architecture.md

```diff
@@ -0,0 +1,685 @@
+
+`state` returns the status of the container. For `kata-runtime`, this means being
+able to detect if the container is still running by looking at the state of `kata-shim`
+process representing this container process.
+
+1. Ask the container status by checking information stored on disk. (clarification needed)
+2. Check `kata-shim` process representing the container.
+3. In case the container status on disk was supposed to be `ready` or `running`,
+ and the `kata-shim` process no longer exists, this involves the detection of a
+ stopped container. This means that before returning the container status,
+ the container has to be properly stopped. Here are the steps involved in this detection:
+	1. Wait for `kata-shim` process to exit.
+	2. Force kill the container process if `kata-shim` process didn't return after a timeout. This is done by communicating with `kata-agent` (connecting the proxy), sending `SIGKILL` signal to the container process inside the VM.
+	3. Wait for `kata-shim` process to exit, and return an error if we reach the timeout again.
+	4. Communicate with `kata-agent` (connecting the proxy) to remove the container configuration from the VM.
+4. Return container status.
+
+## Proxy
+
+Communication with the VM can be achieved by either `virtio-serial` or, if the host
+kernel is newer than v4.8, a virtual socket, `vsock` can be used. The default is `virtio-serial`.
+
+The VM will likely be running multiple container processes.  In the event `virtio-serial`
+is used, the I/O streams associated with each process needs to be multiplexed and demultiplexed on the host. On systems with `vsock` support, this component becomes optional.
+
+`kata-proxy` is a process offering access to the VM [`kata-agent`](https://github.com/kata-containers/agent)
+to multiple `kata-shim` and `kata-runtime` clients associated with the VM. Its
+main role is to route the I/O streams and signals between each `kata-shim`
+instance and the `kata-agent`.
+`kata-proxy` connects to `kata-agent` on a unix domain socket that `kata-runtime` provides
+while spawning `kata-proxy`.
+`kata-proxy` uses [`yamux`](https://github.com/hashicorp/yamux) to multiplex gRPC
+requests on its connection to the `kata-agent`.
+
+When proxy type is configured as "proxyBuiltIn", we do not spawn a separate
+process to proxy grpc connections. Instead a built-in yamux grpc dialer is used to connect
+directly to `kata-agent`. This is used by CRI container runtime server `frakti` which
+calls directly into `kata-runtime`.
+
+## Shim
+
+A container process reaper, such as Docker's `containerd-shim` or CRI-O's `conmon`,
+is designed around the assumption that it can monitor and reap the actual container
+process. As the container process reaper runs on the host, it cannot directly
+monitor a process running within a virtual machine. At most it can see the QEMU
+process, but that is not enough. With Kata Containers, `kata-shim` acts as the
```

Comment:
s/`kata-shim` acts as the container process/a `kata-shim` instance represents the real container process running inside the virtual machine/